File: crates/flow-ingest/src/session.rs
//! Immutable on-disk storage of capture sessions.
//!
//! A session is one directory in the canonical format (ADR 0001):
//!
//! ```text
//! <root>/<session_id>/
//! ├── meta.json        # SessionMeta
//! ├── csi.ndjson       # one CsiFrame per line, append-ordered by ts_us
//! └── labels.ndjson    # one Label per line, append-ordered by ts_us
//! ```
//!
//! While a capture is running, the directory is named
//! `<session_id>.recording`; [`SessionWriter::finalize`] flushes, syncs,
//! and atomically renames it to `<session_id>`. A crash therefore leaves a
//! clearly marked `.recording` directory — a truncated capture can never
//! be mistaken for a clean one.
//!
//! The writer guards the dataset's invariants: structurally invalid
//! frames, frames from nodes not declared in the session metadata, and
//! out-of-order timestamps are all rejected. Recorded sessions are
//! immutable — creating a session whose directory already exists is an
//! error, never an overwrite.

use std::collections::HashSet;
use std::fs::{self, File};
use std::io::{BufRead, BufReader, BufWriter, Write};
use std::path::{Path, PathBuf};

use flow_core::{CsiFrame, FrameError, Label, SessionMeta, TimestampUs};
use thiserror::Error;

/// File name of the session metadata document.
pub const META_FILE: &str = "meta.json";
/// File name of the CSI frame log.
pub const CSI_FILE: &str = "csi.ndjson";
/// File name of the ground-truth label log.
pub const LABELS_FILE: &str = "labels.ndjson";
/// Suffix marking a session directory whose capture is still running
/// (or crashed before finalization).
pub const RECORDING_SUFFIX: &str = ".recording";

/// Failure while creating, writing or reading a session.
#[derive(Debug, Error)]
pub enum SessionError {
    /// The session id is empty or contains characters unsafe for a
    /// directory name.
    #[error("invalid session id: {id:?}")]
    InvalidSessionId {
        /// The offending id.
        id: String,
    },
    /// A directory for this session already exists (recorded sessions are
    /// immutable, and concurrent captures of one id are forbidden).
    #[error("session directory already exists: {}", path.display())]
    AlreadyExists {
        /// The existing path.
        path: PathBuf,
    },
    /// The frame violates the structural invariants of [`CsiFrame`].
    #[error(transparent)]
    Frame(#[from] FrameError),
    /// The frame's node is not declared in the session metadata.
    #[error("frame from node {node_id:?} not declared in session metadata")]
    UnknownNode {
        /// The undeclared node id.
        node_id: String,
    },
    /// A timestamp went backwards within a file (files are append-ordered).
    #[error("out-of-order timestamp: {got} after {last}")]
    OutOfOrder {
        /// Highest timestamp written so far.
        last: TimestampUs,
        /// The offending timestamp.
        got: TimestampUs,
    },
    /// Serialization failure.
    #[error("serialization failed: {0}")]
    Json(#[from] serde_json::Error),
    /// Underlying I/O failure.
    #[error("I/O failure: {0}")]
    Io(#[from] std::io::Error),
}

/// Counts returned by a successful [`SessionWriter::finalize`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SessionSummary {
    /// Final session directory.
    pub path: PathBuf,
    /// Number of CSI frames written.
    pub frames: u64,
    /// Number of labels written.
    pub labels: u64,
}

/// Writer for one capture session in the canonical on-disk format.
///
/// See the module documentation for the directory layout, the
/// `.recording` convention, and the invariants enforced on writes.
#[derive(Debug)]
pub struct SessionWriter {
    recording_dir: PathBuf,
    final_dir: PathBuf,
    csi: BufWriter<File>,
    labels: BufWriter<File>,
    known_nodes: HashSet<String>,
    last_frame_ts: Option<TimestampUs>,
    last_label_ts: Option<TimestampUs>,
    frames: u64,
    label_count: u64,
}

impl SessionWriter {
    /// Creates the session directory under `root` and writes `meta.json`.
    ///
    /// `root` is created if missing. The session directory itself must not
    /// exist yet, in either recording or finalized form.
    ///
    /// # Errors
    ///
    /// [`SessionError::InvalidSessionId`] for an unsafe id,
    /// [`SessionError::AlreadyExists`] if the session exists, or an I/O
    /// error.
    pub fn create(root: &Path, meta: &SessionMeta) -> Result<Self, SessionError> {
        validate_session_id(&meta.session_id)?;
        let final_dir = root.join(&meta.session_id);
        if final_dir.exists() {
            return Err(SessionError::AlreadyExists { path: final_dir });
        }
        let recording_dir = root.join(format!("{}{RECORDING_SUFFIX}", meta.session_id));
        fs::create_dir_all(root)?;
        if let Err(err) = fs::create_dir(&recording_dir) {
            return if err.kind() == std::io::ErrorKind::AlreadyExists {
                Err(SessionError::AlreadyExists {
                    path: recording_dir,
                })
            } else {
                Err(err.into())
            };
        }

        let meta_json = serde_json::to_vec_pretty(meta)?;
        fs::write(recording_dir.join(META_FILE), meta_json)?;

        let csi = BufWriter::new(File::create(recording_dir.join(CSI_FILE))?);
        let labels = BufWriter::new(File::create(recording_dir.join(LABELS_FILE))?);
        let known_nodes = meta.nodes.iter().map(|node| node.node_id.clone()).collect();

        Ok(Self {
            recording_dir,
            final_dir,
            csi,
            labels,
            known_nodes,
            last_frame_ts: None,
            last_label_ts: None,
            frames: 0,
            label_count: 0,
        })
    }

    /// Appends one CSI frame to `csi.ndjson`.
    ///
    /// # Errors
    ///
    /// Rejects structurally invalid frames, frames from nodes not declared
    /// in the session metadata, and timestamps lower than the previous
    /// frame's.
    pub fn write_frame(&mut self, frame: &CsiFrame) -> Result<(), SessionError> {
        frame.validate()?;
        if !self.known_nodes.contains(&frame.node_id) {
            return Err(SessionError::UnknownNode {
                node_id: frame.node_id.clone(),
            });
        }
        check_order(&mut self.last_frame_ts, frame.ts_us)?;
        serde_json::to_writer(&mut self.csi, frame)?;
        self.csi.write_all(b"\n")?;
        self.frames += 1;
        Ok(())
    }

    /// Appends one ground-truth label to `labels.ndjson`.
    ///
    /// # Errors
    ///
    /// Rejects timestamps lower than the previous label's.
    pub fn write_label(&mut self, label: &Label) -> Result<(), SessionError> {
        check_order(&mut self.last_label_ts, label.ts_us)?;
        serde_json::to_writer(&mut self.labels, label)?;
        self.labels.write_all(b"\n")?;
        self.label_count += 1;
        Ok(())
    }

    /// Flushes and syncs all files, then atomically renames the directory
    /// to its final name, sealing the session.
    ///
    /// # Errors
    ///
    /// Any I/O failure; the directory is left in `.recording` state, so a
    /// failed finalization is indistinguishable from a crash — by design.
    pub fn finalize(self) -> Result<SessionSummary, SessionError> {
        for writer in [self.csi, self.labels] {
            let file = writer.into_inner().map_err(std::io::Error::from)?;
            file.sync_all()?;
        }
        fs::rename(&self.recording_dir, &self.final_dir)?;
        Ok(SessionSummary {
            path: self.final_dir,
            frames: self.frames,
            labels: self.label_count,
        })
    }
}

/// Reader for one recorded session, in the format [`SessionWriter`] produces.
///
/// Frames are handed out one at a time rather than collected: a capture runs
/// to tens of megabytes and the appliance this runs on has 512 MB.
#[derive(Debug)]
pub struct SessionReader {
    dir: PathBuf,
    meta: SessionMeta,
}

impl SessionReader {
    /// Opens a session directory and reads its metadata.
    ///
    /// Accepts a sealed session or one still marked `.recording`, since a
    /// capture that never finished is exactly the one worth looking at.
    ///
    /// # Errors
    ///
    /// [`SessionError::Io`] if the directory or `meta.json` cannot be read,
    /// [`SessionError::Json`] if the metadata does not parse.
    pub fn open(dir: &Path) -> Result<Self, SessionError> {
        let meta: SessionMeta = serde_json::from_slice(&fs::read(dir.join(META_FILE))?)?;
        Ok(Self {
            dir: dir.to_path_buf(),
            meta,
        })
    }

    /// What the capture recorded about itself.
    #[must_use]
    pub fn meta(&self) -> &SessionMeta {
        &self.meta
    }

    /// Ids of the receiving nodes, in the order the metadata declares them.
    #[must_use]
    pub fn rx_node_ids(&self) -> Vec<String> {
        let mut ids: Vec<String> = self
            .meta
            .nodes
            .iter()
            .filter(|node| node.role == flow_core::NodeRole::Rx)
            .map(|node| node.node_id.clone())
            .collect();
        ids.sort();
        ids
    }

    /// Every ground-truth label, in timestamp order.
    ///
    /// Read whole: a label per press is a handful of rows, unlike the frames.
    ///
    /// # Errors
    ///
    /// [`SessionError::Io`] or [`SessionError::Json`] on a malformed file.
    pub fn labels(&self) -> Result<Vec<Label>, SessionError> {
        let file = File::open(self.dir.join(LABELS_FILE))?;
        BufReader::new(file)
            .lines()
            .map(|line| Ok(serde_json::from_str(&line?)?))
            .collect()
    }

    /// The CSI frames, in the order they were received.
    ///
    /// # Errors
    ///
    /// [`SessionError::Io`] if the frame log cannot be opened.
    pub fn frames(&self) -> Result<Frames, SessionError> {
        let file = File::open(self.dir.join(CSI_FILE))?;
        Ok(Frames {
            lines: BufReader::new(file).lines(),
        })
    }
}

/// Frames of one session, yielded one at a time.
#[derive(Debug)]
pub struct Frames {
    lines: std::io::Lines<BufReader<File>>,
}

impl Iterator for Frames {
    type Item = Result<CsiFrame, SessionError>;

    fn next(&mut self) -> Option<Self::Item> {
        let line = match self.lines.next()? {
            Ok(line) => line,
            Err(err) => return Some(Err(err.into())),
        };
        Some(read_frame(&line))
    }
}

fn read_frame(line: &str) -> Result<CsiFrame, SessionError> {
    let frame: CsiFrame = serde_json::from_str(line)?;
    // Deserialization does not enforce the type's invariants, and disk is a
    // trust boundary like any other.
    frame.validate()?;
    Ok(frame)
}

fn check_order(last: &mut Option<TimestampUs>, ts_us: TimestampUs) -> Result<(), SessionError> {
    if let Some(prev) = *last {
        if ts_us < prev {
            return Err(SessionError::OutOfOrder {
                last: prev,
                got: ts_us,
            });
        }
    }
    *last = Some(ts_us);
    Ok(())
}

fn validate_session_id(id: &str) -> Result<(), SessionError> {
    let valid = !id.is_empty()
        && id
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_');
    if valid {
        Ok(())
    } else {
        Err(SessionError::InvalidSessionId { id: id.to_owned() })
    }
}

#[cfg(test)]
mod tests {
    use flow_core::{ClassMapping, CsiFrame, DensityClass, NodePlacement, NodeRole};

    use super::*;

    fn meta(session_id: &str) -> SessionMeta {
        SessionMeta {
            session_id: session_id.to_owned(),
            site: "lab-a".into(),
            environment: "test bench".into(),
            wifi_channel: 6,
            nodes: vec![NodePlacement {
                node_id: "rx-1".into(),
                role: NodeRole::Rx,
                position: "desk".into(),
            }],
            firmware_version: "0.1.0".into(),
            software_version: "0.1.0".into(),
            class_mapping: ClassMapping {
                empty: "0".into(),
                low: "1".into(),
                medium: "2".into(),
                saturated: "3+".into(),
            },
        }
    }

    fn frame(ts_us: u64) -> CsiFrame {
        CsiFrame::new(ts_us, "rx-1", -52, 7, vec![1.0, 2.0], vec![0.0, 0.5]).unwrap()
    }

    fn label(ts_us: u64) -> Label {
        Label {
            ts_us,
            class: DensityClass::Low,
            count: Some(1),
        }
    }

    #[test]
    fn full_session_round_trips_through_disk() {
        let root = tempfile::tempdir().unwrap();
        let meta_in = meta("s-001");
        let mut writer = SessionWriter::create(root.path(), &meta_in).unwrap();
        writer.write_frame(&frame(10)).unwrap();
        writer.write_frame(&frame(20)).unwrap();
        writer.write_label(&label(15)).unwrap();
        let summary = writer.finalize().unwrap();

        assert_eq!(summary.frames, 2);
        assert_eq!(summary.labels, 1);
        assert_eq!(summary.path, root.path().join("s-001"));
        assert!(summary.path.is_dir());
        assert!(!root.path().join("s-001.recording").exists());

        let meta_out: SessionMeta =
            serde_json::from_str(&fs::read_to_string(summary.path.join(META_FILE)).unwrap())
                .unwrap();
        assert_eq!(meta_out, meta_in);

        let csi = fs::read_to_string(summary.path.join(CSI_FILE)).unwrap();
        let frames: Vec<CsiFrame> = csi
            .lines()
            .map(|line| serde_json::from_str(line).unwrap())
            .collect();
        assert_eq!(frames, [frame(10), frame(20)]);

        let labels = fs::read_to_string(summary.path.join(LABELS_FILE)).unwrap();
        let labels: Vec<Label> = labels
            .lines()
            .map(|line| serde_json::from_str(line).unwrap())
            .collect();
        assert_eq!(labels, [label(15)]);
    }

    #[test]
    fn directory_stays_marked_recording_without_finalize() {
        let root = tempfile::tempdir().unwrap();
        let writer = SessionWriter::create(root.path(), &meta("s-002")).unwrap();
        drop(writer);
        assert!(root.path().join("s-002.recording").is_dir());
        assert!(!root.path().join("s-002").exists());
    }

    #[test]
    fn refuses_existing_session() {
        let root = tempfile::tempdir().unwrap();
        let writer = SessionWriter::create(root.path(), &meta("s-003")).unwrap();
        // Same id while recording.
        assert!(matches!(
            SessionWriter::create(root.path(), &meta("s-003")),
            Err(SessionError::AlreadyExists { .. })
        ));
        writer.finalize().unwrap();
        // Same id once finalized: recorded sessions are immutable.
        assert!(matches!(
            SessionWriter::create(root.path(), &meta("s-003")),
            Err(SessionError::AlreadyExists { .. })
        ));
    }

    #[test]
    fn refuses_unsafe_session_ids() {
        let root = tempfile::tempdir().unwrap();
        for id in ["", "a/b", "../escape", "a b", "é"] {
            assert!(
                matches!(
                    SessionWriter::create(root.path(), &meta(id)),
                    Err(SessionError::InvalidSessionId { .. })
                ),
                "id accepted: {id:?}"
            );
        }
    }

    #[test]
    fn refuses_frame_from_undeclared_node() {
        let root = tempfile::tempdir().unwrap();
        let mut writer = SessionWriter::create(root.path(), &meta("s-004")).unwrap();
        let foreign = CsiFrame::new(10, "rx-9", -52, 7, vec![1.0], vec![0.0]).unwrap();
        assert!(matches!(
            writer.write_frame(&foreign),
            Err(SessionError::UnknownNode { node_id }) if node_id == "rx-9"
        ));
    }

    #[test]
    fn refuses_structurally_invalid_frame() {
        let root = tempfile::tempdir().unwrap();
        let mut writer = SessionWriter::create(root.path(), &meta("s-005")).unwrap();
        let mut bad = frame(10);
        bad.len = 99;
        assert!(matches!(
            writer.write_frame(&bad),
            Err(SessionError::Frame(_))
        ));
    }

    #[test]
    fn refuses_out_of_order_timestamps() {
        let root = tempfile::tempdir().unwrap();
        let mut writer = SessionWriter::create(root.path(), &meta("s-006")).unwrap();
        writer.write_frame(&frame(20)).unwrap();
        assert!(matches!(
            writer.write_frame(&frame(10)),
            Err(SessionError::OutOfOrder { last: 20, got: 10 })
        ));
        // Equal timestamps are fine (two nodes, same edge stamp).
        writer.write_frame(&frame(20)).unwrap();

        writer.write_label(&label(5)).unwrap();
        assert!(matches!(
            writer.write_label(&label(4)),
            Err(SessionError::OutOfOrder { last: 5, got: 4 })
        ));
    }

    #[test]
    fn empty_session_finalizes_cleanly() {
        let root = tempfile::tempdir().unwrap();
        let writer = SessionWriter::create(root.path(), &meta("s-007")).unwrap();
        let summary = writer.finalize().unwrap();
        assert_eq!(summary.frames, 0);
        assert_eq!(summary.labels, 0);
        assert!(summary.path.join(CSI_FILE).exists());
    }

    /// A sealed session on disk, ready to be read back.
    fn recorded(root: &Path, id: &str, frames: &[CsiFrame], labels: &[Label]) -> PathBuf {
        let mut writer = SessionWriter::create(root, &meta(id)).unwrap();
        for frame in frames {
            writer.write_frame(frame).unwrap();
        }
        for label in labels {
            writer.write_label(label).unwrap();
        }
        writer.finalize().unwrap().path
    }

    #[test]
    fn what_the_writer_sealed_is_what_the_reader_hands_back() {
        let root = tempfile::tempdir().unwrap();
        let dir = recorded(root.path(), "s-100", &[frame(10), frame(20)], &[label(15)]);

        let reader = SessionReader::open(&dir).unwrap();

        assert_eq!(reader.meta().session_id, "s-100");
        assert_eq!(reader.rx_node_ids(), ["rx-1"]);
        assert_eq!(reader.labels().unwrap(), vec![label(15)]);
        let frames: Vec<CsiFrame> = reader.frames().unwrap().map(Result::unwrap).collect();
        assert_eq!(frames, vec![frame(10), frame(20)]);
    }

    #[test]
    fn a_capture_that_never_finished_can_still_be_read() {
        // The one most worth looking at: something went wrong during it.
        let root = tempfile::tempdir().unwrap();
        let mut writer = SessionWriter::create(root.path(), &meta("s-101")).unwrap();
        writer.write_frame(&frame(10)).unwrap();
        drop(writer);

        let reader = SessionReader::open(&root.path().join("s-101.recording")).unwrap();

        assert_eq!(reader.meta().session_id, "s-101");
        assert_eq!(reader.frames().unwrap().count(), 1);
    }

    #[test]
    fn a_frame_whose_invariants_broke_on_disk_is_reported_not_returned() {
        // Deserialization alone would hand back a frame declaring two
        // subcarriers and carrying one, which every consumer would then index
        // out of bounds.
        let root = tempfile::tempdir().unwrap();
        let dir = recorded(root.path(), "s-102", &[frame(10)], &[]);
        let corrupt =
            r#"{"ts_us":20,"node_id":"rx-1","rssi":-52,"mcs":7,"len":2,"amp":[1.0],"phase":[0.0]}"#;
        let mut file = fs::OpenOptions::new()
            .append(true)
            .open(dir.join(CSI_FILE))
            .unwrap();
        writeln!(file, "{corrupt}").unwrap();
        drop(file);

        let reader = SessionReader::open(&dir).unwrap();
        let outcomes: Vec<_> = reader.frames().unwrap().collect();

        assert!(outcomes[0].is_ok());
        assert!(matches!(outcomes[1], Err(SessionError::Frame(_))));
    }

    #[test]
    fn a_truncated_line_is_reported_where_it_sits() {
        // A crash mid-write leaves half a line. Everything before it is still
        // good data, so the reader reports rather than refuses the session.
        let root = tempfile::tempdir().unwrap();
        let dir = recorded(root.path(), "s-103", &[frame(10)], &[]);
        let mut file = fs::OpenOptions::new()
            .append(true)
            .open(dir.join(CSI_FILE))
            .unwrap();
        write!(file, r#"{{"ts_us":20,"node_id":"rx-1","#).unwrap();
        drop(file);

        let reader = SessionReader::open(&dir).unwrap();
        let outcomes: Vec<_> = reader.frames().unwrap().collect();

        assert_eq!(outcomes.len(), 2);
        assert!(outcomes[0].is_ok());
        assert!(matches!(outcomes[1], Err(SessionError::Json(_))));
    }

    #[test]
    fn a_directory_that_is_not_a_session_is_refused_on_open() {
        let root = tempfile::tempdir().unwrap();
        assert!(matches!(
            SessionReader::open(&root.path().join("nothing-here")),
            Err(SessionError::Io(_))
        ));
    }
}

File: ml/flow_ml/session.py
"""Loading and validation of canonical capture sessions.

This module is the Python mirror of the Rust ``flow-core`` data model. The
on-disk session format (``meta.json``, ``csi.ndjson``, ``labels.ndjson``)
is the contract between the edge (Rust, writes) and the ML pipeline
(Python, reads); field names and encodings are frozen, and both sides pin
them with tests on the same canonical examples.

Everything read from disk crosses a trust boundary, so the loader applies
the same validations as the Rust side: declared subcarrier counts,
timestamp ordering, class range, and node declarations.
"""

from __future__ import annotations

import json
import tarfile
import tempfile
from dataclasses import dataclass
from enum import IntEnum
from pathlib import Path
from typing import cast

META_FILE = "meta.json"
CSI_FILE = "csi.ndjson"
LABELS_FILE = "labels.ndjson"

type JsonObj = dict[str, object]


class SessionFormatError(ValueError):
    """A session file violates the canonical format."""


class DensityClass(IntEnum):
    """Frozen 4-class output space of the density classifier.

    The integer encoding (0..3) is part of the on-disk format and must
    never change. The order follows increasing density, which downstream
    smoothing relies on.
    """

    EMPTY = 0
    LOW = 1
    MEDIUM = 2
    SATURATED = 3


@dataclass(frozen=True, slots=True)
class Frame:
    """One CSI measurement — one line of ``csi.ndjson``."""

    ts_us: int
    node_id: str
    rssi: int
    mcs: int
    amp: tuple[float, ...]
    phase: tuple[float, ...]


@dataclass(frozen=True, slots=True)
class Label:
    """One ground-truth annotation — one line of ``labels.ndjson``.

    ``density`` maps to the JSON key ``class`` (a Python keyword).
    ``count`` is the exact people count when a reference sensor measured
    it, ``None`` for manually produced labels.
    """

    ts_us: int
    density: DensityClass
    count: int | None = None


@dataclass(frozen=True, slots=True)
class NodePlacement:
    """Physical placement of one sensing node."""

    node_id: str
    role: str
    position: str


@dataclass(frozen=True, slots=True)
class SessionMeta:
    """Session metadata — ``meta.json``."""

    session_id: str
    site: str
    environment: str
    wifi_channel: int
    nodes: tuple[NodePlacement, ...]
    firmware_version: str
    software_version: str
    class_mapping: dict[DensityClass, str]


@dataclass(frozen=True, slots=True)
class Session:
    """One loaded capture session (immutable, ordered)."""

    meta: SessionMeta
    frames: tuple[Frame, ...]
    labels: tuple[Label, ...]

    @property
    def rx_node_ids(self) -> tuple[str, ...]:
        """Ids of the receiving nodes, in stable (sorted) order."""
        return tuple(sorted(n.node_id for n in self.meta.nodes if n.role == "rx"))

    def frames_of(self, node_id: str) -> tuple[Frame, ...]:
        """Frames received by one node, in timestamp order."""
        return tuple(f for f in self.frames if f.node_id == node_id)


def load_session(path: Path) -> Session:
    """Loads and validates one session directory.

    Raises :class:`SessionFormatError` on any deviation from the canonical
    format — a session that loads is safe to consume downstream.
    """
    meta = _parse_meta(_read_json(path / META_FILE))
    node_ids = {n.node_id for n in meta.nodes}

    frames: list[Frame] = []
    for where, obj in _ndjson_objects(path / CSI_FILE):
        frame = _parse_frame(obj, where)
        if frame.node_id not in node_ids:
            raise SessionFormatError(f"{where}: undeclared node {frame.node_id!r}")
        if frames and frame.ts_us < frames[-1].ts_us:
            raise SessionFormatError(f"{where}: out-of-order timestamp {frame.ts_us}")
        frames.append(frame)

    labels: list[Label] = []
    for where, obj in _ndjson_objects(path / LABELS_FILE):
        label = _parse_label(obj, where)
        if labels and label.ts_us < labels[-1].ts_us:
            raise SessionFormatError(f"{where}: out-of-order timestamp {label.ts_us}")
        labels.append(label)

    return Session(meta=meta, frames=tuple(frames), labels=tuple(labels))


def load_sessions(root: Path) -> list[Session]:
    """Loads every sealed session under `root`, sorted by name.

    Accepts both shapes a capture arrives in: a session directory, and the
    gzipped tar the appliance exports, whose members sit under the session
    identifier.
    """
    sessions: list[Session] = []
    for child in sorted(root.iterdir()):
        if child.is_dir() and not child.name.endswith(".recording"):
            if (child / META_FILE).exists():
                sessions.append(load_session(child))
        elif child.name.endswith(".tar.gz"):
            sessions.append(_load_archive(child))
    return sessions


def _load_archive(path: Path) -> Session:
    with tempfile.TemporaryDirectory() as staging:
        with tarfile.open(path, "r:gz") as archive:
            archive.extractall(staging, filter="data")
        found = [d for d in Path(staging).iterdir() if (d / META_FILE).exists()]
        if len(found) != 1:
            raise SessionFormatError(
                f"{path}: expected one session directory in the archive, found {len(found)}"
            )
        return load_session(found[0])


def _read_json(path: Path) -> JsonObj:
    try:
        parsed: object = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise SessionFormatError(f"{path.name}: invalid JSON: {exc}") from exc
    if not isinstance(parsed, dict):
        raise SessionFormatError(f"{path.name}: expected a JSON object")
    return cast(JsonObj, parsed)


def _ndjson_objects(path: Path) -> list[tuple[str, JsonObj]]:
    out: list[tuple[str, JsonObj]] = []
    for lineno, line in enumerate(path.read_text(encoding="utf-8").splitlines(), start=1):
        if not line.strip():
            continue
        where = f"{path.name}:{lineno}"
        try:
            parsed: object = json.loads(line)
        except json.JSONDecodeError as exc:
            raise SessionFormatError(f"{where}: invalid JSON: {exc}") from exc
        if not isinstance(parsed, dict):
            raise SessionFormatError(f"{where}: expected a JSON object")
        out.append((where, cast(JsonObj, parsed)))
    return out


def _int(obj: JsonObj, key: str, where: str) -> int:
    value = obj.get(key)
    # In Python, bool is a subclass of int: reject it explicitly so that
    # `"rssi": true` does not silently pass as 1.
    if isinstance(value, bool) or not isinstance(value, int):
        raise SessionFormatError(f"{where}: field {key!r} must be an integer")
    return value


def _str(obj: JsonObj, key: str, where: str) -> str:
    value = obj.get(key)
    if not isinstance(value, str):
        raise SessionFormatError(f"{where}: field {key!r} must be a string")
    return value


def _float_seq(obj: JsonObj, key: str, where: str) -> tuple[float, ...]:
    value = obj.get(key)
    if not isinstance(value, list):
        raise SessionFormatError(f"{where}: field {key!r} must be an array")
    out: list[float] = []
    for item in cast(list[object], value):
        if isinstance(item, bool) or not isinstance(item, int | float):
            raise SessionFormatError(f"{where}: field {key!r} must contain numbers")
        out.append(float(item))
    return tuple(out)


def _parse_frame(obj: JsonObj, where: str) -> Frame:
    amp = _float_seq(obj, "amp", where)
    phase = _float_seq(obj, "phase", where)
    declared = _int(obj, "len", where)
    if len(amp) != len(phase):
        raise SessionFormatError(f"{where}: amp/phase length mismatch ({len(amp)}/{len(phase)})")
    if declared != len(amp):
        raise SessionFormatError(f"{where}: declared len {declared} does not match {len(amp)}")
    if not amp:
        raise SessionFormatError(f"{where}: frame carries no subcarrier data")
    return Frame(
        ts_us=_int(obj, "ts_us", where),
        node_id=_str(obj, "node_id", where),
        rssi=_int(obj, "rssi", where),
        mcs=_int(obj, "mcs", where),
        amp=amp,
        phase=phase,
    )


def _parse_label(obj: JsonObj, where: str) -> Label:
    raw_class = _int(obj, "class", where)
    try:
        density = DensityClass(raw_class)
    except ValueError as exc:
        raise SessionFormatError(f"{where}: invalid density class {raw_class}") from exc
    count = _int(obj, "count", where) if "count" in obj else None
    return Label(ts_us=_int(obj, "ts_us", where), density=density, count=count)


def _parse_meta(obj: JsonObj) -> SessionMeta:
    where = META_FILE

    raw_nodes = obj.get("nodes")
    if not isinstance(raw_nodes, list):
        raise SessionFormatError(f"{where}: field 'nodes' must be an array")
    nodes: list[NodePlacement] = []
    for raw_node in cast(list[object], raw_nodes):
        if not isinstance(raw_node, dict):
            raise SessionFormatError(f"{where}: each node must be an object")
        raw = cast(JsonObj, raw_node)
        role = _str(raw, "role", where)
        if role not in ("tx", "rx"):
            raise SessionFormatError(f"{where}: invalid node role {role!r}")
        nodes.append(
            NodePlacement(
                node_id=_str(raw, "node_id", where),
                role=role,
                position=_str(raw, "position", where),
            )
        )

    raw_mapping = obj.get("class_mapping")
    if not isinstance(raw_mapping, dict):
        raise SessionFormatError(f"{where}: field 'class_mapping' must be an object")
    mapping_obj = cast(JsonObj, raw_mapping)
    mapping: dict[DensityClass, str] = {}
    for density in DensityClass:
        mapping[density] = _str(mapping_obj, density.name.lower(), where)

    return SessionMeta(
        session_id=_str(obj, "session_id", where),
        site=_str(obj, "site", where),
        environment=_str(obj, "environment", where),
        wifi_channel=_int(obj, "wifi_channel", where),
        nodes=tuple(nodes),
        firmware_version=_str(obj, "firmware_version", where),
        software_version=_str(obj, "software_version", where),
        class_mapping=mapping,
    )
